proofs/proof_bullet.ml

```diff
@@ -131,11 +131,11 @@ module Strict = struct
            *under this* bullet b, see if a new b can be pushed. *)
           begin
             try ignore (push b prf); Suggest b
-            with _ ->
+            with e when CErrors.noncritical e ->
               (* b could not be pushed, so we must look for a outer bullet *)
               loop prf
           end
-        | exception _ ->
+        | exception e when CErrors.noncritical e ->
```

Comment:
Regarding the test failure in `proof_bullet.ml`: It appears that this used to catch an `assert false` coming from `pop` above. That seems really fragile programming to me. But I don't fully understand what the intended semantics is. Should `pop` return an `Option` instead of crashing? (According to the docs, `pop` can also raise `Proof.CannotUnfocusThisWay`, so it is not clear what this exception is intending to catch.)

@SkySkimmer you seem to be the last person to clean up this code in 2017. Could you take a look at how this is supposed to work?

Reply:
It seems it used to raise ExhaustedBulletStack but was incorrectly changed to assert false.

Reply:
Indeed. But `ExhausedBulletStack` no longer exists now. I'm generally uncomfortable with all this use of exceptions for local control flow. So I rewrite this part to use `option`.